File: apps/web/test/fixtures/inventory.ts
import type {
  InventoryLocationResponse,
  InventoryResponse,
  InventorySummaryResponse,
  PageInventoryResponse,
  PageStockMovementResponse,
  PageStockReservationResponse,
  StockMovementResponse,
  StockReservationResponse,
  StockValuationResponse,
  TransferReportResponse,
  UnitResponse,
  WasteReportResponse,
} from "@/lib/api/inventory";

/** Shared fixtures for 7I Inventory tests -- mirrors test/fixtures/plants.ts's pattern. */

export function makeUnit(overrides: Partial<UnitResponse> = {}): UnitResponse {
  return {
    id: "unit-each-01",
    code: "each",
    name: "Each",
    unit_type: "count",
    ...overrides,
  };
}

export function makeInventoryLocation(overrides: Partial<InventoryLocationResponse> = {}): InventoryLocationResponse {
  return {
    id: "loc-11111111-1111-1111-1111-111111111101",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: "44444444-4444-4444-4444-444444444444",
    parent_location_id: null,
    location_type: "greenhouse",
    name: "Greenhouse 1",
    code: "GH1",
    is_active: true,
    created_at: "2026-01-01T00:00:00Z",
    updated_at: "2026-01-01T00:00:00Z",
    ...overrides,
  };
}

export function makeInventoryItem(overrides: Partial<InventoryResponse> = {}): InventoryResponse {
  return {
    id: "inv-99999999-9999-9999-9999-999999999901",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: "44444444-4444-4444-4444-444444444444",
    species_id: null,
    category_id: "cat-houseplant-01",
    unit_id: "unit-each-01",
    location_id: "loc-11111111-1111-1111-1111-111111111101",
    name: "4in nursery pots",
    quantity: 100,
    reserved_quantity: 10,
    damaged_quantity: 2,
    disposed_quantity: 0,
    available_quantity: 88,
    unit_cost: 0.5,
    unit_price: 1.25,
    low_stock_threshold: 20,
    archived_at: null,
    version: 1,
    created_at: "2026-01-01T00:00:00Z",
    updated_at: "2026-08-01T00:00:00Z",
    ...overrides,
  };
}

export function makeInventoryPage(items: InventoryResponse[] = [makeInventoryItem()]): PageInventoryResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeStockMovement(overrides: Partial<StockMovementResponse> = {}): StockMovementResponse {
  return {
    id: "mv-aaaaaaaa-1111-1111-1111-111111111101",
    inventory_id: makeInventoryItem().id,
    movement_type: "incoming",
    quantity_delta: 50,
    quantity_after: 100,
    reason: null,
    from_location_id: null,
    to_location_id: "loc-11111111-1111-1111-1111-111111111101",
    plant_id: null,
    reservation_id: null,
    transfer_group_id: null,
    reference_sale_id: null,
    reference_purchase_order_id: null,
    note: "Initial receipt",
    performed_by_user_id: "11111111-1111-1111-1111-111111111111",
    created_at: "2026-08-01T00:00:00Z",
    ...overrides,
  };
}

export function makeStockMovementPage(items: StockMovementResponse[] = [makeStockMovement()]): PageStockMovementResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeStockReservation(overrides: Partial<StockReservationResponse> = {}): StockReservationResponse {
  return {
    id: "res-bbbbbbbb-2222-2222-2222-222222222201",
    nursery_id: "22222222-2222-2222-2222-222222222222",
    branch_id: "44444444-4444-4444-4444-444444444444",
    inventory_id: makeInventoryItem().id,
    quantity: 10,
    status: "active",
    reference_type: "sales_order",
    reference_id: null,
    reserved_by_user_id: "11111111-1111-1111-1111-111111111111",
    reserved_at: "2026-08-01T00:00:00Z",
    released_at: null,
    expires_at: null,
    note: null,
    ...overrides,
  };
}

export function makeStockReservationPage(items: StockReservationResponse[] = [makeStockReservation()]): PageStockReservationResponse {
  return { items, meta: { page: 1, page_size: 20, total_items: items.length, total_pages: 1 } };
}

export function makeInventorySummary(overrides: Partial<InventorySummaryResponse> = {}): InventorySummaryResponse {
  return {
    line_count: 5,
    total_quantity: 500,
    total_reserved_quantity: 20,
    total_damaged_quantity: 5,
    total_disposed_quantity: 0,
    total_available_quantity: 475,
    low_stock_count: 1,
    total_valuation: 250.0,
    ...overrides,
  };
}

export function makeStockValuation(overrides: Partial<StockValuationResponse> = {}): StockValuationResponse {
  return {
    line_count: 5,
    total_cost_value: 250.0,
    total_retail_value: 625.0,
    potential_margin: 375.0,
    ...overrides,
  };
}

export function makeWasteReport(overrides: Partial<WasteReportResponse> = {}): WasteReportResponse {
  return {
    movement_count: 0,
    total_quantity_disposed: 0,
    movements: [],
    ...overrides,
  };
}

export function makeTransferReport(overrides: Partial<TransferReportResponse> = {}): TransferReportResponse {
  return {
    movement_count: 0,
    movements: [],
    ...overrides,
  };
}

File: apps/api/app/schemas/inventory.py
"""Pydantic request/response DTOs for Module 8 (Inventory & Stock Management)."""
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.db.enums import InventoryAdjustmentReason, InventoryLocationType, StockMovementType, StockReservationStatus


# ------------------------------------------------------------------
# InventoryLocation
# ------------------------------------------------------------------


class CreateInventoryLocationRequest(BaseModel):
    branch_id: uuid.UUID
    location_type: InventoryLocationType
    name: str = Field(..., max_length=255)
    code: str | None = Field(None, max_length=50)
    parent_location_id: uuid.UUID | None = None


class InventoryLocationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    parent_location_id: uuid.UUID | None
    location_type: InventoryLocationType
    name: str
    code: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class UnitResponse(BaseModel):
    """
    Global, system-seeded reference data (migration 0002) -- the unit of
    measure dropdown `CreateInventoryLineRequest.unit_id` needs. See
    `GET /units`'s summary and `UnitRepository`'s docstring for why this
    exists (a real gap found while building 7I: unlike `category_id`,
    `unit_id` had no route a caller could use to discover valid ids).
    """

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    code: str
    name: str
    unit_type: str


# ------------------------------------------------------------------
# Inventory
# ------------------------------------------------------------------


class CreateInventoryLineRequest(BaseModel):
    branch_id: uuid.UUID
    category_id: uuid.UUID
    unit_id: uuid.UUID
    name: str = Field(..., max_length=255)
    species_id: uuid.UUID | None = None
    location_id: uuid.UUID | None = None
    unit_cost: float | None = Field(None, ge=0)
    unit_price: float | None = Field(None, ge=0)
    low_stock_threshold: int = Field(10, ge=0)
    initial_quantity: int = Field(0, ge=0)


class InventoryResponse(BaseModel):
    """
    "Real-Time Stock": Current/Reserved/Damaged/Disposed are real,
    persisted `Inventory` columns; `available_quantity` (= quantity -
    reserved - damaged) is derived, not stored (avoiding a fifth mutable
    column that could drift out of sync with the other four) -- computed
    once here by `inventory_response()` rather than left to every route
    to recompute individually.
    """

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    species_id: uuid.UUID | None
    category_id: uuid.UUID
    unit_id: uuid.UUID
    location_id: uuid.UUID | None
    name: str
    quantity: int
    reserved_quantity: int
    damaged_quantity: int
    disposed_quantity: int
    available_quantity: int
    unit_cost: float | None
    unit_price: float | None
    low_stock_threshold: int
    archived_at: datetime | None
    version: int
    created_at: datetime
    updated_at: datetime


def inventory_response(item) -> InventoryResponse:
    data = {
        "id": item.id, "nursery_id": item.nursery_id, "branch_id": item.branch_id,
        "species_id": item.species_id, "category_id": item.category_id, "unit_id": item.unit_id,
        "location_id": item.location_id, "name": item.name, "quantity": item.quantity,
        "reserved_quantity": item.reserved_quantity, "damaged_quantity": item.damaged_quantity,
        "disposed_quantity": item.disposed_quantity,
        "available_quantity": item.quantity - item.reserved_quantity - item.damaged_quantity,
        "unit_cost": item.unit_cost, "unit_price": item.unit_price,
        "low_stock_threshold": item.low_stock_threshold, "archived_at": item.archived_at,
        "version": item.version, "created_at": item.created_at, "updated_at": item.updated_at,
    }
    return InventoryResponse(**data)


class ReceiveStockRequest(BaseModel):
    quantity: int = Field(..., gt=0)
    to_location_id: uuid.UUID | None = None
    reference_purchase_order_id: uuid.UUID | None = None
    note: str | None = Field(None, max_length=1000)


class TransferStockRequest(BaseModel):
    quantity: int = Field(..., gt=0)
    to_location_id: uuid.UUID | None = None
    to_branch_id: uuid.UUID | None = None
    note: str | None = Field(None, max_length=1000)


class AdjustStockRequest(BaseModel):
    quantity_delta: int = Field(..., description="Signed -- positive adds stock, negative removes it. Cannot be zero.")
    reason: InventoryAdjustmentReason
    note: str | None = Field(None, max_length=1000)


class MarkDamagedRequest(BaseModel):
    quantity: int = Field(..., gt=0)
    note: str | None = Field(None, max_length=1000)


class DisposeStockRequest(BaseModel):
    quantity: int = Field(..., gt=0)
    from_damaged: bool = False
    plant_id: uuid.UUID | None = None
    note: str | None = Field(None, max_length=1000)


class SellStockRequest(BaseModel):
    quantity: int = Field(..., gt=0)
    reference_sale_id: uuid.UUID | None = None
    plant_id: uuid.UUID | None = None


class ArchiveInventoryRequest(BaseModel):
    reason: str | None = Field(None, max_length=500)


# ------------------------------------------------------------------
# StockMovement
# ------------------------------------------------------------------


class StockMovementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    inventory_id: uuid.UUID
    movement_type: StockMovementType
    quantity_delta: int
    quantity_after: int
    reason: InventoryAdjustmentReason | None
    from_location_id: uuid.UUID | None
    to_location_id: uuid.UUID | None
    plant_id: uuid.UUID | None
    reservation_id: uuid.UUID | None
    transfer_group_id: uuid.UUID | None
    reference_sale_id: uuid.UUID | None
    reference_purchase_order_id: uuid.UUID | None
    note: str | None
    performed_by_user_id: uuid.UUID | None
    created_at: datetime


# ------------------------------------------------------------------
# StockReservation
# ------------------------------------------------------------------


class ReserveStockRequest(BaseModel):
    quantity: int = Field(..., gt=0)
    reference_type: str | None = Field(None, max_length=50)
    reference_id: uuid.UUID | None = None
    expires_at: datetime | None = None
    note: str | None = Field(None, max_length=1000)


class FulfillReservationRequest(BaseModel):
    reference_sale_id: uuid.UUID | None = None
    plant_id: uuid.UUID | None = None


class StockReservationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    nursery_id: uuid.UUID
    branch_id: uuid.UUID
    inventory_id: uuid.UUID
    quantity: int
    status: StockReservationStatus
    reference_type: str | None
    reference_id: uuid.UUID | None
    reserved_by_user_id: uuid.UUID | None
    reserved_at: datetime
    released_at: datetime | None
    expires_at: datetime | None
    note: str | None


# ------------------------------------------------------------------
# Reporting
# ------------------------------------------------------------------


class InventorySummaryResponse(BaseModel):
    line_count: int
    total_quantity: int
    total_reserved_quantity: int
    total_damaged_quantity: int
    total_disposed_quantity: int
    total_available_quantity: int
    low_stock_count: int
    total_valuation: float


class WasteReportResponse(BaseModel):
    movement_count: int
    total_quantity_disposed: int
    movements: list[StockMovementResponse]


class TransferReportResponse(BaseModel):
    movement_count: int
    movements: list[StockMovementResponse]


class StockValuationResponse(BaseModel):
    line_count: int
    total_cost_value: float
    total_retail_value: float
    potential_margin: float
